### ISSUE TITLE
Neglected memory cache

### ISSUE THREAD

**Issue by mforde84** [alexdobin/STAR#381]
Hi Alex,

I'm sure you must have run into his before, but we Have a production HPC environment and people  will occasionally load a genome to memory, but forget to remove it following completion of their alignments. So after a while we run into memory issues on these nodes, and have to hard reboot. I was thinking we could simply clear the cache in a cronjob, but this might cause issues with legitimate running jobs. Just curious if you have any suggestions or could point me in the right direction.

Thanks,
Martin

**Comment by alexdobin** [comment]
Hi Martin,

sorry for the belayed reply. In principle, it's safe to remove shared memory with ipcrm - if there is still a job attached to it, it will only remove it after the job finishes. However, my recommendation would be to run a cronjob that would check if any jobs are attached to the shared memory block with ipcs - if no jobs are attached, remove it. This will prevent removing the shared memory in the middle of a sequential batch of jobs.

Cheers
Alex

**Comment by NathanielMiddleton** [comment]
Depending on the scheduler being used, an epilogue script addition would be a cleaner way to accomplish this (vs. cron). I rolled this into our environment:
if ! [[ $(ipcs -m | grep -q ^0x ) ]]
then
        ipcs -m | grep ${USER::10} | awk '{if ($6 == "0") system( "ipcrm " "-M" $1);}'
fi

While there is an issue of duplicate users on the first 10 characters, I ended up keeping this shorter as nattach = 0 is a fairly good way of sorting unused shmem.

I am not sure if any schedulers are actually keeping track of IPCs, so be aware that jobs may be scheduled to consume the full node of ram alongside the STAR jobs (thus running OOM)

Wanted to add some extra thoughts- in hopes it will help people.

-Nathaniel
